README.md
# shared-biubiu-crypto-utils


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.6](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/compare/v1.0.5...v1.0.6) (2023-10-13)


### Bug Fixes

* **fix:** d ([3a0519e](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/commit/3a0519eb4c69fbc3542b0ea59c60768724cdbe49))
* **fx:** ff ([4e18407](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/commit/4e18407cb11cafa3612e55b269177246b0c13089))

### [1.0.5](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/compare/v1.0.4...v1.0.5) (2023-10-13)


### Bug Fixes

* **ttt:** ttt ([f424262](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/commit/f4242621764d3f8c66756e385bb9b2b7180aaebf))

### [1.0.1](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/compare/v1.0.4...v1.0.1) (2023-10-13)


### Bug Fixes

* **ttt:** ttt ([f424262](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/commit/f4242621764d3f8c66756e385bb9b2b7180aaebf))

### [1.0.4](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/compare/v1.0.3...v1.0.4) (2023-10-13)


### Bug Fixes

* **aa:** aaad ([addb83b](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/commit/addb83b749b5278537a85962cc413a422ae2d6c3))
* **fa:** sdf ([4dc6921](https://github.com/YOUR_GITHUB_USER_NAME/shared-biubiu-crypto-lib/commit/4dc69213f0a9a7cc95f6ea6afd8b38925557163d))


src/tests/utils.spec.ts
import test from 'ava';
import { bytesToString, stringToBytes } from 'viem';

import {
  decrypt,
  encrypt,
  hexToTronBs58,
  toAptosAddress,
  toAptosWIF,
  toBTCAddress,
  toBTCWIF,
  toETHAddress,
  toETHWIF,
  toSolanaAddress,
  toSolanaWIF,
  toSUIAddress,
  toSUIWIF,
  toTronAddress,
  toTronWIF,
  tronBs58ToHex,
} from '../lib/utils.js';

test('toETHAddress-valid-65', async (t) => {
  const pubkey = new Uint8Array([
    4, 132, 70, 215, 81, 0, 182, 70, 225, 201, 109, 114, 64, 13, 78, 237, 150,
    17, 118, 207, 169, 57, 69, 29, 232, 32, 74, 37, 98, 85, 57, 35, 36, 142,
    166, 99, 190, 190, 234, 111, 66, 58, 135, 82, 83, 163, 131, 192, 54, 211,
    36, 245, 222, 224, 237, 45, 240, 51, 39, 151, 171, 215, 226, 93, 180,
  ]);
  const address = await toETHAddress(pubkey);

  t.is(address.toLowerCase(), '0xd3a13dfc223cb67c8f1882f1fa17ea9daa776959');
});

test('toETHAddress-valid-64', async (t) => {
  const pubkey = new Uint8Array([
    132, 70, 215, 81, 0, 182, 70, 225, 201, 109, 114, 64, 13, 78, 237, 150, 17,
    118, 207, 169, 57, 69, 29, 232, 32, 74, 37, 98, 85, 57, 35, 36, 142, 166,
    99, 190, 190, 234, 111, 66, 58, 135, 82, 83, 163, 131, 192, 54, 211, 36,
    245, 222, 224, 237, 45, 240, 51, 39, 151, 171, 215, 226, 93, 180,
  ]);
  const address = await toETHAddress(pubkey);
  t.is(address.toLowerCase(), '0xd3a13dfc223cb67c8f1882f1fa17ea9daa776959');
});

test('toETHAddress-invalid', async (t) => {
  const pubkey = new Uint8Array([
    70, 215, 81, 0, 182, 70, 225, 201, 109, 114, 64, 13, 78, 237, 150, 17, 118,
    207, 169, 57, 69, 29, 232, 32, 74, 37, 98, 85, 57, 35, 36, 142, 166, 99,
    190, 190, 234, 111, 66, 58, 135, 82, 83, 163, 131, 192, 54, 211, 36, 245,
    222, 224, 237, 45, 240, 51, 39, 151, 171, 215, 226, 93, 180,
  ]);

  try {
    await toETHAddress(pubkey);
  } catch (error) {
    t.is(typeof error, 'object');
  }
});

test('toTronAddress', async (t) => {
  const pubkey = new Uint8Array([
    132, 70, 215, 81, 0, 182, 70, 225, 201, 109, 114, 64, 13, 78, 237, 150, 17,
    118, 207, 169, 57, 69, 29, 232, 32, 74, 37, 98, 85, 57, 35, 36, 142, 166,
    99, 190, 190, 234, 111, 66, 58, 135, 82, 83, 163, 131, 192, 54, 211, 36,
    245, 222, 224, 237, 45, 240, 51, 39, 151, 171, 215, 226, 93, 180,
  ]);
  const address = await toTronAddress(pubkey);

  t.is(address, 'TVGCjqNEHhu2WHk5YS48r4A5x3tfEzKGac');
});

test('tronBs58ToHex', (t) => {
  t.is(
    tronBs58ToHex('TVGCjqNEHhu2WHk5YS48r4A5x3tfEzKGac').toLowerCase(),
    '0xd3a13dfc223cb67c8f1882f1fa17ea9daa776959',
  );
});

test('hexToTronBs58', async (t) => {
  t.is(
    await hexToTronBs58('0xd3a13dfc223cb67c8f1882f1fa17ea9daa776959'),
    'TVGCjqNEHhu2WHk5YS48r4A5x3tfEzKGac',
  );
});

test('toBTCAddress-P2PKH', async (t) => {
  const pubkey = new Uint8Array([
    2, 20, 170, 58, 33, 247, 51, 151, 138, 244, 104, 175, 56, 1, 230, 234, 193,
    52, 115, 3, 113, 115, 177, 25, 24, 16, 24, 148, 62, 103, 214, 80, 183,
  ]);
  t.is(
    await toBTCAddress(pubkey, 'P2PKH'),
    '13Rqm5AyPqs3KhpGYWACNfeYtq8D4Ak63e',
  );
});

test('toBTCAddress-P2PKH-another-pubkey', async (t) => {
  const pubkey = new Uint8Array([
    2, 180, 19, 234, 67, 117, 25, 97, 221, 118, 210, 23, 30, 146, 218, 212, 26,
    114, 253, 249, 111, 61, 159, 1, 118, 58, 43, 56, 168, 232, 56, 37, 18,
  ]);

  t.is(
    await toBTCAddress(pubkey, 'P2PKH'),
    '1PvSuRvaqBPnzKg39qCGzPVXKx6CHFZ29H',
  );
});

test('toBTCAddress-P2WPKH', async (t) => {
  const pubkey = new Uint8Array([
    2, 180, 19, 234, 67, 117, 25, 97, 221, 118, 210, 23, 30, 146, 218, 212, 26,
    114, 253, 249, 111, 61, 159, 1, 118, 58, 43, 56, 168, 232, 56, 37, 18,
  ]);

  t.is(
    await toBTCAddress(pubkey, 'P2WPKH'),
    'bc1qldhxlnen3xqyw25k3vtcrjne2gph5anm4mptl9',
  );
});

test('toBTCAddress-P2SH_P2WPKH', async (t) => {
  const pubkey = new Uint8Array([
    2, 180, 19, 234, 67, 117, 25, 97, 221, 118, 210, 23, 30, 146, 218, 212, 26,
    114, 253, 249, 111, 61, 159, 1, 118, 58, 43, 56, 168, 232, 56, 37, 18,
  ]);

  t.is(
    await toBTCAddress(pubkey, 'P2SH-P2WPKH'),
    '38pVaiGCcX5ccWhyPjTsibCXx2MCb1TPDq',
  );
});

test('toBTCAddress-P2SH_P2WPKH-another-pubkey', async (t) => {
  const pubkey = new Uint8Array([
    2, 63, 179, 216, 245, 117, 19, 113, 233, 207, 58, 30, 146, 190, 20, 61, 221,
    75, 11, 183, 31, 20, 177, 37, 21, 144, 42, 4, 226, 153, 185, 82, 145,
  ]);

  t.is(
    await toBTCAddress(pubkey, 'P2SH-P2WPKH'),
    '3CMGKR3hepw4CsrxEunQh351jYvwhRCUkX',
  );
});

test('toBTCWIF', async (t) => {
  const privkey = new Uint8Array([
    2, 0, 120, 14, 111, 191, 217, 240, 218, 105, 87, 100, 166, 39, 247, 251, 26,
    195, 249, 125, 17, 132, 70, 130, 187, 205, 15, 184, 241, 250, 110, 74,
  ]);

  t.is(
    await toBTCWIF(privkey),
    'KwHbszP8MiLRCuq7D1yT7sDKxFnS4X2nzUDEWayBh3KzTSBgAChH',
  );
});
test('toETHWIF', async (t) => {
  const privkey = new Uint8Array([
    4, 147, 16, 65, 254, 46, 35, 232, 253, 155, 121, 241, 109, 48, 46, 161, 229,
    180, 195, 94, 254, 227, 226, 2, 12, 187, 16, 252, 183, 200, 187, 115,
  ]);

  t.is(
    await toETHWIF(privkey),
    '0x04931041fe2e23e8fd9b79f16d302ea1e5b4c35efee3e2020cbb10fcb7c8bb73',
  );
});

test('toTronWIF', async (t) => {
  const privkey = new Uint8Array([
    4, 147, 16, 65, 254, 46, 35, 232, 253, 155, 121, 241, 109, 48, 46, 161, 229,
    180, 195, 94, 254, 227, 226, 2, 12, 187, 16, 252, 183, 200, 187, 115,
  ]);

  t.is(
    await toTronWIF(privkey),
    '04931041fe2e23e8fd9b79f16d302ea1e5b4c35efee3e2020cbb10fcb7c8bb73',
  );
});

test('toAptosWIF', async (t) => {
  const privkey = new Uint8Array([
    138, 59, 150, 235, 43, 19, 81, 92, 192, 5, 214, 170, 33, 1, 79, 249, 115,
    136, 237, 160, 124, 52, 223, 140, 71, 254, 44, 46, 56, 149, 20, 220,
  ]);

  t.is(
    await toAptosWIF(privkey),
    '0x8a3b96eb2b13515cc005d6aa21014ff97388eda07c34df8c47fe2c2e389514dc',
  );
});

test('toAptosAddress', async (t) => {
  const pubkey = new Uint8Array([
    52, 200, 88, 71, 104, 163, 158, 226, 198, 228, 179, 205, 216, 242, 14, 45,
    184, 17, 56, 92, 169, 192, 118, 213, 141, 203, 153, 252, 50, 146, 124, 107,
    0,
  ]);

  t.is(
    await toAptosAddress(pubkey),
    '0xa547c5e4f85513bd6f35199e20cf0f9217816d56781d69a8d72aed79482a7713',
  );
});

test('toSolanaAddress', async (t) => {
  const pubkey = new Uint8Array([
    233, 139, 15, 100, 3, 244, 105, 93, 115, 82, 87, 52, 196, 128, 110, 118, 61,
    66, 9, 218, 50, 51, 124, 239, 16, 167, 211, 55, 227, 164, 232, 228,
  ]);

  t.is(
    await toSolanaAddress(pubkey),
    'GieynY44ruznz3JgRte2PsbA3oV24TLJGtAKMjbMXwRm',
  );
});

test('toSolanaWIF', async (t) => {
  const privkey = new Uint8Array([
    37, 131, 108, 156, 68, 90, 169, 138, 22, 234, 77, 252, 188, 146, 41, 99, 76,
    18, 56, 65, 195, 106, 124, 108, 50, 71, 170, 114, 253, 185, 204, 57,
  ]);

  t.is(
    await toSolanaWIF(privkey),
    'kW3SuswccnXVaxU9muMj9AFHSqg71qfpesyJqNKMNReAHiNwtuWnQPvJ8p6YoS9JiGnVDLWY75DukxBX3T1CgDy',
  );
});

test('toSUIAddress', async (t) => {
  const pubkey = new Uint8Array([
    53, 14, 97, 41, 170, 86, 123, 15, 211, 132, 6, 216, 156, 40, 56, 251, 93,
    204, 211, 230, 110, 221, 217, 27, 25, 170, 66, 60, 134, 157, 202, 102,
  ]);

  t.is(
    await toSUIAddress(pubkey),
    '0xe95bc03686707062944fcd7511ced71ef7ca693079b9f427cf17dccc83c6b096',
  );
});

test('toSUIAddress-another-pubkey', async (t) => {
  const pubkey = new Uint8Array([
    52, 200, 88, 71, 104, 163, 158, 226, 198, 228, 179, 205, 216, 242, 14, 45,
    184, 17, 56, 92, 169, 192, 118, 213, 141, 203, 153, 252, 50, 146, 124, 107,
  ]);

  t.is(
    await toSUIAddress(pubkey),
    '0x90b2dc7e5b36eeb3e754b12decb0de37d569ebe535df1bbb50aef15c63d7257a',
  );
});

test('toSUIWIF', async (t) => {
  const privkey = new Uint8Array([
    138, 59, 150, 235, 43, 19, 81, 92, 192, 5, 214, 170, 33, 1, 79, 249, 115,
    136, 237, 160, 124, 52, 223, 140, 71, 254, 44, 46, 56, 149, 20, 220,
  ]);

  t.is(
    await toSUIWIF(privkey),
    '0x8a3b96eb2b13515cc005d6aa21014ff97388eda07c34df8c47fe2c2e389514dc',
  );
});

test('encrypt-decrypt', async (t) => {
  const key = new Uint8Array([
    3, 253, 71, 39, 90, 227, 158, 75, 46, 142, 84, 140, 88, 6, 87, 97, 127, 77,
    22, 13, 62, 64, 252, 132, 64, 5, 214, 209, 179, 1, 82, 189, 146,
  ]);

  const text = 'Text may be any length you wish, no padding is required.';
  const rawBytes = stringToBytes(text);

  const encryptedBytes = await encrypt(key, rawBytes);

  const decryptedBytes = await decrypt(key, encryptedBytes);
  const decyptedText = bytesToString(decryptedBytes);

  t.is(text, decyptedText);
});


src/tests/Ed25519.spec.ts
import { ed25519 } from '@noble/curves/ed25519';
import test from 'ava';

import Ed25519 from '../lib/Ed25519.js';
import { areUint8ArraysEqual } from '../lib/utils.js';

test('generateKeyPair', (t) => {
  const instance = new Ed25519();
  const res = instance.generateKeyPair();
  t.is(areUint8ArraysEqual(res[1], ed25519.getPublicKey(res[0])), true);
});

test('generateKeyPairFromPrivkey', (t) => {
  const instance = new Ed25519();
  const res = instance.generateKeyPairFromPrivkey(
    new Uint8Array([
      88, 103, 130, 127, 121, 233, 9, 70, 180, 193, 248, 46, 115, 245, 214, 32,
      38, 59, 70, 215, 248, 137, 43, 249, 149, 122, 104, 197, 45, 233, 98, 154,
    ]),
  );

  const pubkey = new Uint8Array([
    147, 248, 141, 123, 247, 249, 239, 33, 8, 76, 9, 226, 247, 180, 182, 217,
    50, 9, 31, 163, 247, 216, 68, 25, 121, 180, 39, 143, 135, 21, 191, 77,
  ]);

  t.is(areUint8ArraysEqual(res[1], pubkey), true);
});

test('generateSharedSecret', (t) => {
  const privatekey1 = new Uint8Array([
    176, 232, 250, 117, 58, 21, 165, 133, 204, 126, 48, 52, 77, 253, 153, 21,
    115, 9, 12, 84, 32, 72, 44, 202, 206, 207, 86, 63, 162, 50, 87, 77,
  ]);
  // x25519 pubkey not ed25519 pubkey
  const pubkey1 = new Uint8Array([
    14, 247, 135, 183, 238, 152, 140, 112, 148, 46, 192, 155, 164, 235, 88, 17,
    137, 185, 191, 205, 104, 240, 36, 172, 1, 124, 27, 38, 251, 203, 184, 0,
  ]);

  const privatekey2 = new Uint8Array([
    88, 103, 130, 127, 121, 233, 9, 70, 180, 193, 248, 46, 115, 245, 214, 32,
    38, 59, 70, 215, 248, 137, 43, 249, 149, 122, 104, 197, 45, 233, 98, 154,
  ]);

  // x25519 pubkey not ed25519 pubkey
  const pubkey2 = new Uint8Array([
    124, 142, 89, 185, 97, 0, 34, 130, 160, 5, 157, 242, 235, 158, 120, 48, 3,
    29, 95, 247, 78, 108, 224, 88, 211, 22, 68, 102, 129, 48, 195, 12,
  ]);

  const instance = new Ed25519();

  const secret1 = instance.generateSharedSecret(pubkey2, privatekey1);
  const secret2 = instance.generateSharedSecret(pubkey1, privatekey2);

  t.is(areUint8ArraysEqual(secret1, secret2), true);
});

test('sign-verify', (t) => {
  const message = new Uint8Array(32).fill(1);
  const privatekey = new Uint8Array([
    88, 103, 130, 127, 121, 233, 9, 70, 180, 193, 248, 46, 115, 245, 214, 32,
    38, 59, 70, 215, 248, 137, 43, 249, 149, 122, 104, 197, 45, 233, 98, 154,
  ]);
  const pubkey = new Uint8Array([
    147, 248, 141, 123, 247, 249, 239, 33, 8, 76, 9, 226, 247, 180, 182, 217,
    50, 9, 31, 163, 247, 216, 68, 25, 121, 180, 39, 143, 135, 21, 191, 77,
  ]);

  const instance = new Ed25519();
  const signature = instance.sign(message, privatekey);

  t.is(instance.verify(signature, message, pubkey), true);
});

test('getHDKey', (t) => {
  const instance = new Ed25519();
  // secret passphrase: 223
  const mnemonic =
    'finger plug meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';
  const HDKey = instance.getHDKey(mnemonic);

  const hdkey = HDKey.derive(`m/44'/637'/0'/0'/5'`);
  const pk = new Uint8Array([
    8, 126, 233, 231, 221, 159, 9, 82, 69, 150, 217, 145, 88, 50, 5, 175, 195,
    157, 71, 84, 10, 122, 229, 48, 98, 196, 99, 80, 16, 217, 119, 68,
  ]);
  t.is(areUint8ArraysEqual(hdkey.privateKey, pk), true);
});

test('deriveFromSeed', (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'finger plug meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';
  const pair = instance.deriveFromSeed(mnemonic, `m/44'/637'/0'/0'/5'`);
  const pk = new Uint8Array([
    8, 126, 233, 231, 221, 159, 9, 82, 69, 150, 217, 145, 88, 50, 5, 175, 195,
    157, 71, 84, 10, 122, 229, 48, 98, 196, 99, 80, 16, 217, 119, 68,
  ]);
  const res = instance.generateKeyPairFromPrivkey(pk);
  t.is(
    areUint8ArraysEqual(pair[0], pk) && areUint8ArraysEqual(pair[1], res[1]),
    true,
  );
});

test('deriveFromHDkey', (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'finger plug meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';
  const HDKey = instance.getHDKey(mnemonic);

  const pair = instance.deriveFromHDkey(HDKey, `m/44'/637'/0'/0'/5'`);
  const pk = new Uint8Array([
    8, 126, 233, 231, 221, 159, 9, 82, 69, 150, 217, 145, 88, 50, 5, 175, 195,
    157, 71, 84, 10, 122, 229, 48, 98, 196, 99, 80, 16, 217, 119, 68,
  ]);

  const res = instance.generateKeyPairFromPrivkey(pk);
  t.is(
    areUint8ArraysEqual(pair[0], pk) && areUint8ArraysEqual(pair[1], res[1]),
    true,
  );
});


src/tests/Cryptolib.spec.ts
import test from 'ava';

import Ed25519 from '../lib/Ed25519.js';

test('generateMnemonic-24', (t) => {
  const instance = new Ed25519();
  const mnemonic = instance.generateMnemonic(24);
  t.is(mnemonic.split(' ').length, 24);
});

test('generateMnemonic-12', (t) => {
  const instance = new Ed25519();
  const mnemonic = instance.generateMnemonic(12);
  t.is(mnemonic.split(' ').length, 12);
});

test('mnemonicToSeedSync', (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'finger plug meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';

  const seed = instance.mnemonicToSeedSync(mnemonic);
  t.is(seed instanceof Uint8Array && seed.length === 64, true);
});
test('validateMnemonic-valid', (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'finger plug meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';

  t.is(instance.validateMnemonic(mnemonic), true);
});
test('validateMnemonic-invalid', (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'plug finger meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';

  t.is(instance.validateMnemonic(mnemonic), false);
});
test('getMnemonic-24', async (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'finger plug meadow swamp filter object second vintage embrace correct knee bullet deposit film fish involve route sustain embark broccoli cable reveal oyster panda';

  const mnemonic2 = await instance.getMnemonic('223', 24);
  t.is(mnemonic2, mnemonic);
});

test('getMnemonic-12', async (t) => {
  const instance = new Ed25519();
  const mnemonic =
    'bacon wisdom favorite give lottery relax mouse trend window typical smooth buddy';

  const mnemonic2 = await instance.getMnemonic('223', 12);
  t.is(mnemonic2, mnemonic);
});


src/lib/Ed25519.ts
import { ed25519, x25519 } from '@noble/curves/ed25519';
import * as bip39 from '@scure/bip39';
import { HDKey } from 'ed25519-keygen/hdkey';
import type { HDKey as HDKeyType } from 'ed25519-keygen/hdkey';

import Cryptolib from './Cryptolib.js';
import type { Ed25519KeyPair, PrivKey, PubKey } from './Interface.js';

export default class Ed25519 extends Cryptolib {
  generateKeyPair(): Ed25519KeyPair {
    const privKey = ed25519.utils.randomPrivateKey();
    const pubKey = ed25519.getPublicKey(privKey);
    const x25519pubkey = x25519.getPublicKey(privKey);

    return [privKey, pubKey, x25519pubkey];
  }

  generateKeyPairFromPrivkey(privateKey: PrivKey): Ed25519KeyPair {
    const pubKey = ed25519.getPublicKey(privateKey);
    const x25519pubkey = x25519.getPublicKey(privateKey);
    return [privateKey, pubKey, x25519pubkey];
  }

  generateSharedSecret(publicKey: PubKey, privateKey: PrivKey): Uint8Array {
    return x25519.getSharedSecret(privateKey, publicKey);
  }

  sign(message: Uint8Array, privateKey: PrivKey): Uint8Array {
    return ed25519.sign(message, privateKey);
  }

  verify(signature: Uint8Array, message: Uint8Array, pubkey: PubKey): boolean {
    return ed25519.verify(signature, message, pubkey);
  }

  // HDKEY
  getHDKey(mnemonicOrXpub: string): HDKeyType {
    const seed = bip39.mnemonicToSeedSync(mnemonicOrXpub);
    return HDKey.fromMasterSeed(seed);
  }

  deriveFromSeed(mnemonic: string, path: string): Ed25519KeyPair {
    const seed = bip39.mnemonicToSeedSync(mnemonic);
    const hdkey = HDKey.fromMasterSeed(seed);
    const deriveNode = hdkey.derive(path);
    const pubKey = ed25519.getPublicKey(deriveNode.privateKey);
    const x25519pubkey = x25519.getPublicKey(deriveNode.privateKey);
    return [deriveNode.privateKey, pubKey, x25519pubkey];
  }

  deriveFromHDkey(hdkey: HDKeyType, path: string): Ed25519KeyPair {
    const deriveNode = hdkey.derive(path);
    const x25519pubkey = x25519.getPublicKey(deriveNode.privateKey);
    const pubKey = ed25519.getPublicKey(deriveNode.privateKey);
    return [deriveNode.privateKey, pubKey, x25519pubkey];
  }
}


src/lib/Cryptolib.ts
import * as bip39 from '@scure/bip39';
import { wordlist } from '@scure/bip39/wordlists/english';
import { md5, sha256 } from 'hash-wasm';
import { hexToBytes } from 'viem';

import type { Ed25519KeyPair, KeyPair, PrivKey, PubKey } from './Interface';
export default abstract class Cryptolib {
  abstract generateKeyPair(): KeyPair | Ed25519KeyPair;
  abstract generateSharedSecret(
    publicKey: PubKey,
    privateKey: PrivKey,
  ): Uint8Array;
  abstract sign(message: Uint8Array, privateKey: PrivKey): Uint8Array;
  abstract verify(
    signature: Uint8Array,
    message: Uint8Array,
    pubkey: PubKey,
  ): boolean;

  // mnemonic
  generateMnemonic(length: number) {
    if (length == 24) {
      return bip39.generateMnemonic(wordlist, 256);
    } else {
      return bip39.generateMnemonic(wordlist, 128);
    }
  }

  mnemonicToSeedSync(mnemonic: string): Uint8Array {
    return bip39.mnemonicToSeedSync(mnemonic);
  }

  validateMnemonic(mnemonic: string): boolean {
    return bip39.validateMnemonic(mnemonic, wordlist);
  }

  async getMnemonic(passphrase: string, length: number): Promise<string> {
    const passphraseTrim = passphrase.trim();
    if (length == 24) {
      const sha256Checksum = await sha256(passphraseTrim);
      return bip39.entropyToMnemonic(
        hexToBytes(`0x${sha256Checksum}`),
        wordlist,
      );
    } else {
      const md5Checksum = await md5(passphraseTrim);
      return bip39.entropyToMnemonic(hexToBytes(`0x${md5Checksum}`), wordlist);
    }
  }
}


src/lib/utils.ts
import aesjs from 'aes-js';
import { BitcoinAddress } from 'bech32-buffer';
import bs58 from 'bs58';
import { blake2b, keccak, ripemd160, sha256, sha3 } from 'hash-wasm';
import { bytesToHex, hexToBytes } from 'viem';
import type { Hex } from 'viem';

import Ed25519 from './Ed25519.js';
import type {
  BTCAddressType,
  ETHAddress,
  PrivKey,
  PubKey,
  TronAddress,
} from './Interface';

// to address
export const toETHAddress = async (
  uncompressedPubKey: PubKey,
): Promise<ETHAddress> => {
  if (uncompressedPubKey.length == 65) {
    const hash = await keccak(uncompressedPubKey.slice(1, 65), 256);
    return `0x${hash.slice(-40)}`;
  }

  if (uncompressedPubKey.length == 64) {
    const hash = await keccak(uncompressedPubKey, 256);
    return `0x${hash.slice(-40)}`;
  }

  throw new Error('Invalid uncompressedPubKey');
};
export const toTronAddress = async (
  uncompressedPubKey: PubKey,
): Promise<TronAddress> => {
  const hexAddr = await toETHAddress(uncompressedPubKey);
  return await hexToTronBs58(hexAddr);
};

export const hexToTronBs58 = async (
  hexAddr: ETHAddress,
): Promise<TronAddress> => {
  const bytes = hexToBytes(hexAddr.replace('0x', '0x41') as Hex);

  const hash0 = await sha256(bytes);
  const hash1 = await sha256(hexToBytes(`0x${hash0}`));

  const checkSum = hexToBytes(`0x${hash1}`).slice(0, 4);

  return bs58.encode(new Uint8Array([...bytes, ...checkSum]));
};

export const tronBs58ToHex = (bs58Addr: TronAddress): ETHAddress => {
  const bytes = bs58.decode(bs58Addr);
  return `0x${bytesToHex(bytes).replace('0x41', '')}`.slice(
    0,
    42,
  ) as ETHAddress;
};

export const toETHWIF = (privkey: PrivKey): string => {
  return bytesToHex(privkey);
};

export const toTronWIF = (privkey: PrivKey): string => {
  return bytesToHex(privkey).replace('0x', '');
};

export const toAptosWIF = (privkey: PrivKey): string => {
  return bytesToHex(privkey.slice(0, 32));
};

export const toSUIWIF = (privkey: PrivKey): string => {
  return bytesToHex(privkey.slice(0, 32));
};

export const toSolanaWIF = (privkey: PrivKey): string => {
  const ed25519 = new Ed25519();
  const keypair = ed25519.generateKeyPairFromPrivkey(privkey);

  return bs58.encode(new Uint8Array([...keypair[0], ...keypair[1]]));
};

export const toBTCWIF = async (privkey: PrivKey): Promise<string> => {
  if (privkey.length == 32) {
    const data = new Uint8Array([128, ...privkey, 1]);
    const hash0 = await sha256(data);
    const hash1 = await sha256(hexToBytes(`0x${hash0}`));

    const checkSum = hexToBytes(`0x${hash1}`).slice(0, 4);

    return bs58.encode(new Uint8Array([...data, ...checkSum]));
  }

  throw new Error('Invalid privateKey length');
};
export const toBTCAddress = async (
  pubkey: PubKey,
  addressType: BTCAddressType,
) => {
  if (pubkey.length == 33) {
    if (addressType == 'P2PKH') {
      return await P2PKH(pubkey);
    }

    if (addressType == 'P2WPKH') {
      return await P2WPKH(pubkey);
    }
    if (addressType == 'P2SH-P2WPKH') {
      const address = await P2WPKH(pubkey);
      return await P2SH_P2WPKH(address);
    }

    throw new Error('Invalid addressType');
  }

  throw new Error('Invalid pubkey');
};

export const hash160 = async (data: Uint8Array): Promise<Uint8Array> => {
  const hash0 = await sha256(data);
  const hash1 = await ripemd160(hexToBytes(`0x${hash0}`));
  return hexToBytes(`0x${hash1}`);
};

export const P2PKH = async (pubkey: PubKey) => {
  const hash0 = await sha256(pubkey);
  const hash1 = await ripemd160(hexToBytes(`0x${hash0}`));
  const bytes0 = hexToBytes(`0x00${hash1}`);
  const hash2 = await sha256(bytes0);
  const hash3 = await sha256(hexToBytes(`0x${hash2}`));
  const checkSum = hexToBytes(`0x${hash3}`).slice(0, 4);

  return bs58.encode(new Uint8Array([...bytes0, ...checkSum]));
};

export const P2WPKH = async (pubkey: PubKey) => {
  const hash0 = await hash160(pubkey);
  const prefix = 'bc';
  const scriptVersion = 0;

  return new BitcoinAddress(prefix, scriptVersion, hash0).encode();
};

export const P2SH_P2WPKH = async (P2WPKHAddr: string) => {
  const decoded = BitcoinAddress.decode(P2WPKHAddr);
  const version = 0;
  const len = 20;
  const redeemScript = new Uint8Array([version, len, ...decoded.data]);
  return await P2SH(redeemScript);
};

export const P2SH = async (pubkey: PubKey) => {
  const hash0 = await hash160(pubkey);
  const versionPrefix = 5;
  const hash1 = await sha256(new Uint8Array([versionPrefix, ...hash0]));
  const hash2 = await sha256(hexToBytes(`0x${hash1}`));
  const checkSum = hexToBytes(`0x${hash2}`).slice(0, 4);
  return bs58.encode(new Uint8Array([versionPrefix, ...hash0, ...checkSum]));
};

export const toAptosAddress = async (pubkey: PubKey): Promise<string> => {
  const hash = await sha3(pubkey, 256);
  return `0x${hash}`;
};
export const toSolanaAddress = (pubkey: PubKey): string => {
  return bs58.encode(pubkey);
};
export const toSUIAddress = async (pubkey: PubKey): Promise<string> => {
  const data = await blake2b(new Uint8Array([0, ...pubkey]), 256);
  return normalizeSuiAddress(data.slice(0, SUI_ADDRESS_LENGTH * 2));
};

// AES encrypt
export const encrypt = async (
  key: Uint8Array,
  rawBytes: Uint8Array,
): Promise<Uint8Array> => {
  if (key.length >= 32) {
    const encryptKey = key.slice(0, 32);
    const aesCtr = new aesjs.ModeOfOperation.ctr(encryptKey);
    const encryptedBytes = aesCtr.encrypt(rawBytes);

    return encryptedBytes;
  }

  throw new Error('Invalid key');
};
export const decrypt = async (
  key: Uint8Array,
  encryptedBytes: Uint8Array,
): Promise<Uint8Array> => {
  if (key.length >= 32) {
    const decryptKey = key.slice(0, 32);
    const aesCtr = new aesjs.ModeOfOperation.ctr(decryptKey);
    const rawBytes = aesCtr.decrypt(encryptedBytes);

    return rawBytes;
  }

  throw new Error('Invalid key');
};

export const areUint8ArraysEqual = (arr1: Uint8Array, arr2: Uint8Array) => {
  if (arr1.length !== arr2.length) {
    return false;
  }

  return arr1.every((value, index) => value === arr2[index]);
};

export const SUI_ADDRESS_LENGTH = 32;
export function normalizeSuiAddress(
  value: string,
  forceAdd0x: boolean = false,
): string {
  let address = value.toLowerCase();
  if (!forceAdd0x && address.startsWith('0x')) {
    address = address.slice(2);
  }
  return `0x${address.padStart(SUI_ADDRESS_LENGTH * 2, '0')}`;
}
